ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/batch/BatchedSubmissionValidator.scala

```diff
@@ -0,0 +1,439 @@
+      implicit executionContext: ExecutionContext): Source[Stage1, NotUsed] =
+    Source(
+      Indexed
+        .fromSeq(batch.getSubmissionsList.asScala
+          .map(cs => cs.getCorrelationId -> cs.getSubmission))
+        .to)
+      .mapAsyncUnordered(params.cpuParallelism) {
+        _.mapFuture {
+          case (correlationId, submissionEnvelope) =>
+            // Decompress and decode the submissions in parallel.
+            Timed.timedAndTrackedFuture(
+              Metrics.decode,
+              Metrics.decodeRunning,
+              Future {
+                val submission = Envelope
+                  .openSubmission(submissionEnvelope)
+                  .fold(error => throw validator.ValidationFailed.ValidationError(error), identity)
+                Metrics.receivedSubmissionBytes.update(submission.getSerializedSize)
+                CorrelatedSubmission(
+                  correlationId,
+                  bytesToLogEntryId(submissionEnvelope),
+                  submission)
+              }
+            )
+        }
+      }
+
+  private type DamlInputState = Map[DamlStateKey, Option[DamlStateValue]]
+
+  //
+  // The following type aliases describe how the batch processing pipeline transforms the data:
+  //
+
+  // The batch pipeline starts with a stream of correlated submissions that carry their original index
+  // in the batch.
+  private type Stage1 = Indexed[CorrelatedSubmission]
+
+  // The second stage resolves the inputs to each submission.
+  private type FetchedInput = (CorrelatedSubmission, DamlInputState)
+  private type Stage2 = Indexed[FetchedInput]
+
+  // Third stage validates the submission, adding in the validation results.
+  private type ValidatedSubmission = (CorrelatedSubmission, DamlInputState, LogEntryAndState)
+  private type Stage3 = Indexed[ValidatedSubmission]
+
+  // Fourth stage collects the results.
+  private type Stage4 = List[Stage3]
+
+  // The fifth stage sorts the results and drops the index.
+  private type Stage5 = ValidatedSubmission
+
+  // Sixth stage performs conflict detection and potentially drops conflicting results.
+  private type Stage6 = Stage5
+
+  // The last stage commits the results.
+  private type Stage7 = Unit
+
+  /** Validate and commit a batch of indexed DAML submissions.
+    * See the type definitions above to understand the different stages in the
+    * processing pipeline.
+    */
+  private def processBatch(
+      participantId: ParticipantId,
+      batchCorrelationId: CorrelationId,
+      recordTime: Timestamp,
+      indexedSubmissions: Source[Stage1, NotUsed],
+      damlLedgerStateReader: DamlLedgerStateReader,
+      commitStrategy: CommitStrategy[CommitResult])(
+      implicit materializer: Materializer,
+      executionContext: ExecutionContext,
+      logCtx: LoggingContext): Future[Unit] =
+    indexedSubmissions
+    /** Stage1 => Stage2: Fetch the submission inputs in parallel. */
```

Comment:
Pertaining to the above, I'd call this "Stage 1", which goes from "Input" to "Output 1" (or something like that).

Reply:
Removed the stage references from here -- now the comments are easier to read.